hosted-ledgers/shop-with-hosted-ledger.js

```diff
@@ -0,0 +1,151 @@
+
+  // Handle incoming web requests
+  http.createServer(function (req, res) {
+      // Generate a client ID and a shared secret from which this client
+      // can derive fulfillment/condition pairs.
+      const clientId = base64url(crypto.randomBytes(8))
+      const sharedSecret = crypto.randomBytes(32)
+
+      // Store the shared secret and the http request context to use when we get paid
+      sharedSecrets[clientId]  = { sharedSecret, res }
+
+      console.log(`    - Waiting for payments...`)
+
+      res.writeHead(200, {
+        Pay: `interledger-psk ${cost} ${account}.${clientId} ${base64url(sharedSecret)}`
+      })
+      // Flush the headers in a first TCP packet:
+      res.socket.write(res._header)
+      res._headerSent = true
+  }).listen(8000, function () {
+    console.log(`    - Listening on http://localhost:8000`)
+  })
+
+  // Handle incoming payments
+  plugin.on('incoming_prepare', function (transfer) {
+    if (parseInt(transfer.amount) < 10) {
+      // Transfer amount is incorrect
+      console.log(`    - Payment received for the wrong amount ` +
+                                        `(${transfer.amount})... Rejected`)
+
+      const normalizedAmount = transfer.amount /
+                            Math.pow(10, parseInt(ledgerInfo.currencyScale))
+
+      plugin.rejectIncomingTransfer(transfer.id, {
+        code: 'F04',
+        name: 'Insufficient Destination Amount',
+        message: `Please send at least 10 ${ledgerInfo.currencyCode},` +
+                  `you sent ${normalizedAmount}`,
+        triggered_by: plugin.getAccount(),
+        triggered_at: new Date().toISOString(),
+        forwarded_by: [],
+        additional_info: {}
+      })
+      return
+    }
+    // Generate fulfillment from packet and this client's shared secret
+    const ilpPacket = Buffer.from(transfer.ilp, 'base64')
+    const payment = IlpPacket.deserializeIlpPayment(ilpPacket)
+    const clientId = payment.account.substring(plugin.getAccount().length + 1).split('.')[0]
+    const secret = sharedSecrets[clientId].sharedSecret
+    const res = sharedSecrets[clientId].res
+
+    if (!clientId || !secret) {
+      // We don't have a fulfillment for this condition
+      console.log(`    - Payment received with an unknown condition: ` +
+                                            `${transfer.executionCondition}`)
+
+      plugin.rejectIncomingTransfer(transfer.id, {
+        code: 'F05',
+        name: 'Wrong Condition',
+        message: `Unable to fulfill the condition:  ` +
+                                            `${transfer.executionCondition}`,
+        triggered_by: plugin.getAccount(),
+        triggered_at: new Date().toISOString(),
+        forwarded_by: [],
+        additional_info: {}
+      })
+      return
+    }
+    console.log(`    - Calculating hmac; for clientId ${clientId}, the shared secret is ${base64url(secret)}.`)
+    const fulfillmentGenerator = hmac(secret, 'ilp_psk_condition')
```

Comment:
Can you use `ilp`'s PSK?

Reply:
we could just say 'run ilp-kit', but that defeats the purpose of doing a tutorial. in fact, while writing this tutorial, I discovered that the psk spec was not even describing the behavior of the ilp module, so that shows the dangers of depending on one reference implementation and treating it as a black box.